Closes #96 | Create dataset loader for LimeSoda

jen-santoso opened the pull request:
Please name your PR after the issue it closes. You can use the following line: "Closes #96 " where you replace the ISSUE-NUMBER with the one corresponding to your dataset.

### Checkbox
- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `seacrowd/sea_datasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py`.
- [ ] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

ljvmiranda921 commented on seacrowd/sea_datasets/limesoda/limesoda.py:
```diff
@@ -0,0 +1,174 @@
+            return [
+                datasets.SplitGenerator(
+                    name=datasets.Split.TRAIN,
+                    # Whatever you put in gen_kwargs will be passed to _generate_examples
```
```suggestion
```

ljvmiranda921 commented on seacrowd/sea_datasets/limesoda/limesoda.py:
```diff
@@ -0,0 +1,174 @@
+    def _generate_examples(self, filepath: Path) -> Tuple[int, Dict]:
+        with open(filepath, "r") as f:
+            df = [json.loads(line) for line in f.readlines()]
+        f.close()
```
`f.close()`should be unnecessary when you're using the `with` context manager

ljvmiranda921 commented on seacrowd/sea_datasets/limesoda/limesoda.py:
```diff
@@ -0,0 +1,174 @@
+                for i, row in enumerate(df):
+                    ex = {"id": str(i), "text": row["Text"], "document_tag": row["Document Tag"]}
+                    yield i, ex
+        # For example seacrowd_kb, seacrowd_t2t
```
```suggestion
```

ljvmiranda921 commented on seacrowd/sea_datasets/limesoda/limesoda.py:
```diff
@@ -0,0 +1,174 @@
+    DEFAULT_CONFIG_NAME = "limesoda_raw_source"
+
+    def _info(self) -> datasets.DatasetInfo:
+
```
Running `make check_file` requires us to delete this extra newline.

```suggestion
```

ljvmiranda921 commented on seacrowd/sea_datasets/limesoda/limesoda.py:
```diff
@@ -0,0 +1,174 @@
+            return [
+                datasets.SplitGenerator(
+                    name=datasets.Split.TRAIN,
+                    # Whatever you put in gen_kwargs will be passed to _generate_examples
```
```suggestion
```

ljvmiranda921 commented on seacrowd/sea_datasets/limesoda/limesoda.py:
```diff
@@ -0,0 +1,174 @@
+
+    def _generate_examples(self, filepath: Path) -> Tuple[int, Dict]:
+        with open(filepath, "r") as f:
+            df = [json.loads(line) for line in f.readlines()]
```
I suggest changing the variable name `df` into something more informative. I usually see this in the context of a pandas DataFrame, but I don't see any use of that library here.

ljvmiranda921 reviewed (commented):
The dataloader runs as expected, but I have some comments on the source code. Also, the default test errors out. Maybe it's better to rename `limesoda_raw_source` into `limesoda_source`? (and `limesoda_raw_seacrowd_text` as `limesoda_seacrowd_text`). Thoughts @sabilmakbar ?

jen-santoso commented:
> The dataloader runs as expected, but I have some comments on the source code. Also, the default test errors out. Maybe it's better to rename `limesoda_raw_source` into `limesoda_source`? (and `limesoda_raw_seacrowd_text` as `limesoda_seacrowd_text`). Thoughts @sabilmakbar ?

@ljvmiranda921 Thank you for the comments. I was thinking to split the dataset into signify the one with splits and not, therefore naming the splits into `limesoda_raw_source` and `limesoda_split_source`. But think I agree with you with changing `limesoda_raw_source` to `limesoda_source`, it keeps the default runnable and it can still be easily differentiated with the splitted one.
Fixed this, and also the other comments. Please check.

sabilmakbar commented:
> The dataloader runs as expected, but I have some comments on the source code. Also, the default test errors out. Maybe it's better to rename limesoda_raw_source into limesoda_source? (and limesoda_raw_seacrowd_text as limesoda_seacrowd_text). Thoughts @sabilmakbar ?

Hi LJ @ljvmiranda921, actually for several cases, it's okay not to implement the default config of `f"{DATASET_NAME}_source"` and its SEACrowd counterpart, as long as it's justified (belongs to either Multilingual, or have different labels per config and/or has each annotator data provided; for manually curated data -- CebuaNER was one of that case). But esp for the second case, if we can be assured of one default implementation among different configs, it can be defined on `f"{DATASET_NAME}_source"`, which I agree on the fix by @jen-santoso based on ur suggestion.

> @ljvmiranda921 Thank you for the comments. I was thinking to split the dataset into signify the one with splits and not, therefore naming the splits into limesoda_raw_source and limesoda_split_source. But think I agree with you with changing limesoda_raw_source to limesoda_source, it keeps the default runnable and it can still be easily differentiated with the splitted one.
Fixed this, and also the other comments. Please check.

Ack, will verify the dataloader & its metadata in the next 24hrs. But before that, do you mind re-running your dataloader w/ this [Code Formatter Check](https://github.com/SEACrowd/seacrowd-datahub/blob/master/DATALOADER.md#5-format-your-code), if possible? I saw some lines aren't coherently formatted; I suspect the code hasn't been linted using SEACrowd Code Formatter.

ljvmiranda921 commented:
^Ah thanks for that reminder! I'll be more lenient on the the `_source` implementations next time.

ljvmiranda921 reviewed (approved):
LGTM! Let's just wait for @sabilmakbar 's comments

jen-santoso commented:
> 
> Ack, will verify the dataloader & its metadata in the next 24hrs. But before that, do you mind re-running your dataloader w/ this [Code Formatter Check](https://github.com/SEACrowd/seacrowd-datahub/blob/master/DATALOADER.md#5-format-your-code), if possible? I saw some lines aren't coherently formatted; I suspect the code hasn't been linted using SEACrowd Code Formatter.

![image](https://github.com/SEACrowd/seacrowd-datahub/assets/62987563/b308dd9d-40d3-4109-a85e-21041c52b014)
make check_file=seacrowd/sea_datasets/limesoda/limesoda.py result from the latest commit

sabilmakbar reviewed (changes requested):
Two more nits, then all good!

sabilmakbar commented on seacrowd/sea_datasets/limesoda/limesoda.py:
```diff
@@ -0,0 +1,169 @@
+
+    BUILDER_CONFIGS = [
+        SEACrowdConfig(
+            name="limesoda_source",
+            version=SOURCE_VERSION,
+            description="limesoda source schema",
+            schema="source",
+            subset_id="limesoda",
+        ),
+        SEACrowdConfig(
+            name="limesoda_split_source",
+            version=SOURCE_VERSION,
+            description="limesoda source schema",
+            schema="source",
+            subset_id="limesoda_split",
+        ),
+        SEACrowdConfig(
+            name="limesoda_seacrowd_text",
+            version=SEACROWD_VERSION,
+            description="limesoda SEACrowd schema",
+            schema="seacrowd_text",
+            subset_id="limesoda",
+        ),
+        SEACrowdConfig(
+            name="limesoda_split_seacrowd_text",
+            version=SEACROWD_VERSION,
+            description="limesoda: split SEACrowd schema",
+            schema="seacrowd_text",
+            subset_id="limesoda_split",
+        ),
+    ]
+
+    DEFAULT_CONFIG_NAME = "limesoda_source"
```
let's use `_DATASETNAME` variable
```suggestion
            name=f"{_DATASETNAME}_source",
            version=SOURCE_VERSION,
            description="limesoda source schema",
            schema="source",
            subset_id=_DATASETNAME,
        ),
        SEACrowdConfig(
            name=f"{_DATASETNAME}_split_source",
            version=SOURCE_VERSION,
            description="limesoda source schema",
            schema="source",
            subset_id=f"{_DATASETNAME}_split",
        ),
        SEACrowdConfig(
            name=f"{_DATASETNAME}_seacrowd_text",
            version=SEACROWD_VERSION,
            description="limesoda SEACrowd schema",
            schema="seacrowd_text",
            subset_id=_DATASETNAME,
        ),
        SEACrowdConfig(
            name=f"{_DATASETNAME}_split_seacrowd_text",
            version=SEACROWD_VERSION,
            description="limesoda: split SEACrowd schema",
            schema="seacrowd_text",
            subset_id=f"{_DATASETNAME}_split",
        ),
    ]

    DEFAULT_CONFIG_NAME = f"{_DATASETNAME}_source"
```

jen-santoso commented:
Ok, I've updated the nits suggested. Also, I added the `_LANGUAGES` constant as well to ensure consistency as recommended in https://github.com/SEACrowd/seacrowd-datahub/pull/72

jen-santoso commented:
sorry I missed the class name case nits, fixed now.

sabilmakbar commented on seacrowd/sea_datasets/limesoda/limesoda.py:
```diff
@@ -0,0 +1,171 @@
+
+_LICENSE = Licenses.CC_BY_4_0.value
+
+_LANGUAGES = ["tha"]
+
```
can you add constant `_LOCAL = False`? it indicates this dataloader doesn't need an indirect download (which can't be catered by HF `DownloadManager` methods).
```suggestion
_LANGUAGES = ["tha"]
_LOCAL = False

```

sabilmakbar reviewed (commented):
add `_LOCAL` variable to indicate whether this dataset needs to have indirect download or free to be downloaded w/o any gate (HF allows the dataset owner to collect user info prior to accessing dataset, which called Free upon Request dataset or Gated Dataset)

sabilmakbar reviewed (approved):
LGTM! Thanks @jen-santoso for implementing this dataloader and @ljvmiranda921 for reviewing this!
